zebra-state/src/service.rs

```diff
@@ -918,17 +927,10 @@ impl Service<Request> for StateService {
                 let res: Vec<_> = res
                     .iter()
                     .map(|&hash| {
-                        let block = self
-                            .best_block(hash.into())
-                            .expect("block for found hash is in the best chain");
-                        block::CountedHeader {
-                            transaction_count: block
-                                .transactions
-                                .len()
-                                .try_into()
-                                .expect("transaction count has already been validated"),
-                            header: block.header,
-                        }
+                        let header = self
+                            .best_block_header(hash.into())
+                            .expect("block header for found hash is in the best chain");
+                        block::CountedHeader { header }
```

Comment:
This is the core of the bug fix: get headers from the database, rather than entire blocks.